OpenRA.Mods.Common/Traits/Buildings/Dock.cs

```diff
@@ -0,0 +1,125 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public abstract class DockInfo : TraitInfo
+	{
+		[Desc("Actual harvester facing when docking.")]
+		public readonly WAngle DockAngle = WAngle.Zero;
+
+		[Desc("Docking cell relative to top-left cell.")]
+		public readonly CVec DockOffset = CVec.Zero;
+
+		[Desc("Does the refinery require the harvester to be dragged in?")]
+		public readonly bool IsDragRequired = false;
+
+		[Desc("Vector by which the harvester will be dragged when docking.")]
+		public readonly WVec DragOffset = WVec.Zero;
+
+		[Desc("In how many steps to perform the dragging?")]
+		public readonly int DragLength = 0;
+	}
+
+	// TODO: turn this into a trait
+	public abstract class Dock : ITick, INotifySold, INotifyCapture, INotifyOwnerChanged, ISync, INotifyActorDisposing
+	{
+		protected readonly Actor Self;
+		readonly DockInfo info;
+
+		[Sync]
+		Actor dockedHarv = null;
+
+		[Sync]
+		bool preventDock = false;
+
+		public bool AllowDocking => !preventDock;
+		public CVec DeliveryOffset => info.DockOffset;
+		public WAngle DeliveryAngle => info.DockAngle;
+		public bool IsDragRequired => info.IsDragRequired;
+		public WVec DragOffset => info.DragOffset;
+		public int DragLength => info.DragLength;
+
+		public Dock(Actor self, DockInfo info)
+		{
+			Self = self;
+			this.info = info;
+		}
+
+		public abstract Activity DockSequence(Actor harv, Actor self);
+
+		public IEnumerable<TraitPair<Harvester>> GetLinkedHarvesters()
+		{
+			return Self.World.ActorsWithTrait<Harvester>().Where(a => a.Trait.LinkedProc == Self);
```

Comment:
This clearly belongs into `Refinery` and not an abstract class with the generic name `Dock`.

Reply:
It appears so because the DockableBase trait wasn't created yet. Also Dock will become a normal Trait, not an abstract class

Reply:
Also the linking part will get completely rewritten. I'll do that as soon as this PR (and another I don't recall atm) get merged

Reply:
Can you create a pull request that does not do all those intermediate steps?

Reply:
I'm making one at this very moment. It'll abstract DockHost from Refinery, DockClientBase from Harvester, refactor linking and add DockHostManager trait. It's probably take a bit until it's ready to be PR'ed though.

All of the other PR's can still be merged though as they either deal with different areas of the code or add missing features